lib/search_logic.rb

```diff
@@ -0,0 +1,133 @@
+module SearchLogic
+    class << self
+        def get_places_from_city_string(city_name, place_type)
+            begin
+                city = get_city_from_name(city_name)
+                places = city.public_send(place_type).to_a #city.attractions.to_a
+                return places if places.count == 20
+
+                case place_type
+                when "attractions"
+                    category = "tourism.attraction"
+                    response = get_place_details(category, city.city_place_id)
+                    create_attractions(response, city)
+                when "hotels"
+                    category = "accommodation.hotel"
+                    response = get_place_details(category, city.city_place_id)
+                    create_hotels(response, city)
+                when "restaurants"
+                    category = "catering.restaurant"
+                    response = get_place_details(category, city.city_place_id)
+                    create_restaurants(response, city)
+                end
+            rescue
+                raise PlaceTypeInvalid.new("Error: The place type used is not supported at the moment")
+            end
+        end
+
+        private
+
+        def get_city_from_name(city_name)
+            city = City.find_by(name: city_name)
+            return city if city
+
+            response = get_geocode_response(city_name)
+            city_place_id = extract_city_place_id(response)
+            name = extract_city_name(response)
+            city = City.find_or_create_by(city_place_id:)
+            city.name = name
+            city.save
```

Comment:
I forgot one thing in my code snippet:

```suggestion
            city.save!
```

Otherwise, it would proceed with silent save errors, and that could lead to obscure bug instead of obvious one.